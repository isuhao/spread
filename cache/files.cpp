#include "files.hpp"

#include <boost/filesystem.hpp>

using namespace Cache;
using namespace Spread;
namespace bf = boost::filesystem;

std::string Files::makePath(const Hash &hash) const
{
  assert(hash.isSet());
  std::string hstr = hash.toString();
  assert(hstr.size() > 2);
  bf::path res = basedir;
  res /= hstr.substr(0,2);
  res /= hstr.substr(2,std::string::npos);
  return res.string();
}

std::string Files::storePath(const Hash &hash) const
{
  std::string path = makePath(hash);
  bf::create_directories(bf::path(path).parent_path());

  Hash fhash = index.checkFile(path);

  if(fhash != hash)
    // Remove existing file if there is a mismatch
    bf::remove(path);
  else
    // Otherwise, return an empty string, signalling that this file is
    // already OK.
    path = "";

  return path;
}
