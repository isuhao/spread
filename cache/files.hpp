#ifndef __SPREAD_CACHE_FILES_HPP_
#define __SPREAD_CACHE_FILES_HPP_

#include "iindex.hpp"

namespace Cache
{
  struct Files
  {
    Files(ICacheIndex &_index, const std::string &_dir = "")
      : basedir(_dir), index(_index) {}

    /* Set this (directly or through the constructor) before using the
       object.
     */
    std::string basedir;

    /* Produce cache path for a given hash. This does not read or
       write the filesystem, it just creates the path string.
     */
    std::string makePath(const Spread::Hash &hash) const;

    /* Get a path ready for storing files. This will check if the file
       already exists, with the right hash. If it does, the function
       returns "". If a string is returned, the file is guaranteed not
       to exist.

       All necessary parent directories will be created.
     */
    std::string storePath(const Spread::Hash &hash) const;

  private:
    ICacheIndex &index;
  };
}

#endif
