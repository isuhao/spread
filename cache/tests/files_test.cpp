#include <iostream>

#include "index.hpp"
#include "files.hpp"

#include <boost/filesystem.hpp>
#include <mangle/stream/servers/outfile_stream.hpp>

using namespace Spread;
using namespace std;
namespace bf = boost::filesystem;

Cache::CacheIndex cache;
Cache::Files files(cache, "_files");

void test(const Hash &hash)
{
  cout << "HASH: " << hash << endl;
  cout << "Path: " << files.makePath(hash) << endl;

  std::string path = files.storePath(hash);
  cout << "Storing to " << path << endl;
  Mangle::Stream::OutFileStream::Write(path, "hello");
  cout << "Storing again: " << files.storePath(hash) << endl;
  cout << "        twice: " << files.storePath(hash) << endl;
}

int main()
{
  bf::remove_all("_files");

  test(Hash("abcdef"));
  test(Hash("hello",5));

  return 0;
}
