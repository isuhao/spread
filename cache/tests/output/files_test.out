HASH: abcdef
Path: _files/ab/cdef
Storing to _files/ab/cdef
Storing again: _files/ab/cdef
        twice: _files/ab/cdef
HASH: LPJNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
Path: _files/LP/JNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
Storing to _files/LP/JNul-wow4m6DsqxbninhsWHlwfp0JecwQzYpOLmCQF
Storing again: 
        twice: 
